### PR TITLE
feat: add kafka source list command

### LINKED ISSUE #17: Add 'kn source kafka list' command
Add `list` command to kafka source plugin

### PR DESCRIPTION
**Pull request by Kaustubh-pande**
## Description

- Add Kafka source list command
- bump `kn-source-pkg` to v0.6.1
- Add missing test cases for flag factory functions & boilerplate functions

## Reference

Fixes #17

### CODE REVIEW

**Comment by dsimansk** [comment]
Thanks!
/ok-to-test

File: pkg/client/client.go [hunk 1]
```diff
@@ -95,6 +95,16 @@ func (c *kafkaSourceClient) GetKafkaSource(name string) (*v1alpha1.KafkaSource,
 	return kafkaSource, nil
 }
 
+//GetKafkaSources is used to get all available instance of KafkaSource
+func (c *kafkaSourceClient) GetKafkaSources() (*v1alpha1.KafkaSourceList, error) {
```
Comment (dsimansk) [inline pkg/client/client.go:99]: Can you take a look at those `Codecov` annotations and add a few tests to make a bit happier? The client should be pretty straightforward. :)

**Review by maximilien — APPROVED** [review]
Nice work. Add more tests to achieve better coverage and address some of the missing pluralization and duplicated “Description” items.

Thanks for this contribution 🙏🏽

File: pkg/client/client_test.go [hunk 2]
```diff
@@ -17,34 +17,46 @@ package client
...
 	"gotest.tools/assert"
+	"k8s.io/apimachinery/pkg/runtime"
 	v1alpha1 "knative.dev/eventing-kafka/pkg/apis/sources/v1alpha1"
+	"knative.dev/eventing-kafka/pkg/client/clientset/versioned/typed/sources/v1alpha1/fake"
+	"knative.dev/kn-plugin-source-kafka/pkg/types"
 )
 
+var testNamespace = "fake-namespace"
+
+func setup() types.KafkaSourceClient {
```
Comment (dsimansk) [inline pkg/client/client_test.go:31]: It might be a bit cleaner to return both `FakeSourcesV1alpha1, KafkaSourceCLient` as we have it in `knative/client` tests.
E.g.: https://github.com/knative/client/blob/master/pkg/sources/v1alpha2/ping_client_test.go#L30-L35

File: pkg/client/client_test.go [hunk 3]
```diff
@@ -63,6 +75,20 @@ func TestCreateKafkaMultipleTopicsServers(t *testing.T) {
 	assert.NilError(t, err)
 }
 
+func TestGetKafkaSources(t *testing.T) {
+	fakeClient := fake.FakeSourcesV1alpha1{Fake: &client_testing.Fake{}}
```
Comment (dsimansk) [inline pkg/client/client_test.go:79]: WIth updated `setup` function you'd get both vars `fakeClient, cli` initialized.

**Comment by maximilien** [comment]
Hi, @dsimansk, any updates on this? Ready for final review?

File: pkg/factories/rune_factory.go [hunk 4]
```diff
@@ -242,3 +297,41 @@ func writeKafkaSource(dw printers.PrefixWriter, source *v1alpha1.KafkaSource) {
 	dw.WriteAttribute("Topics", strings.Join(source.Spec.Topics, ","))
 	dw.WriteAttribute("ConsumerGroup", source.Spec.ConsumerGroup)
 }
+
+// printKafkaSource populates a single row of kafka source list table
+func printKafkaSource(kafkaSource *v1alpha1.KafkaSource, options printers.PrintOptions) ([]metav1.TableRow, error) {
```
Comment (dsimansk) [inline pkg/factories/rune_factory.go:302]: @Kaustubh-pande I think we can safely remove error from return statement here. And below in the `printKafkaSourceList` function as well. It's not actually passing any error message from the code, only `nil` value.

**Comment by dsimansk** [comment]
Hi, @maximilien sorry I was off for a couple of days. I think yes, it's ready. 

We chatted a bit with @Kaustubh-pande about `codecov` report and test coverage, but it should be fine with the recent commits to meet at least minimal threshold to pass the check. With the current test setup it's difficult to get decent test for e.g. `rune_factory.go`. I'd suggest to revisit the test improvements later to get the list command in first. Wdyt?

File: pkg/factories/rune_factory.go (hunk 4 above)
Comment (Kaustubh-pande) [reply pkg/factories/rune_factory.go]: @dsimansk we need to add back two Err call, inside tableHandler is validates print handler signature & It must be of the following type:
``func printFunc(object ObjectType, options PrintOptions) ([]metav1beta1.TableRow, error)``

**Review by dsimansk — APPROVED** [review]

File: pkg/types/interfaces.go [hunk 5]
```diff
@@ -27,6 +27,7 @@ type KafkaSourceClient interface {
 	CreateKafkaSource(kafkaSource *v1alpha1.KafkaSource) error
 	DeleteKafkaSource(name string) error
 	GetKafkaSource(name string) (*v1alpha1.KafkaSource, error)
+	GetKafkaSources() (*v1alpha1.KafkaSourceList, error)
```
Comment (rhuss) [inline pkg/types/interfaces.go:30]: For the client API we used `ListKafkaSources` and I think it would make sense to you the same naming conventions here, also to avoid confusion with the singular versions of this method.

File: pkg/factories/rune_factory.go [hunk 6]
```diff
@@ -220,6 +223,58 @@ func (f *kafkaSourceRunEFactory) DescribeRunE() sourcetypes.RunE {
...
+		}
+		if len(kafkaSourceList.Items) == 0 {
+			return fmt.Errorf("no kafka source found in namespace '%s'", f.kafkaSourceClient.Namespace())
+		}
+
+		printer := printers.NewTableGenerator()
+		kafkaSourceListColumnDefinitions := []metav1.TableColumnDefinition{
+			{Name: "Name", Type: "string", Description: "Name of the created kafka source", Priority: 1},
+			{Name: "Age", Type: "string", Description: "Age of the kafka source", Priority: 1},
+			{Name: "BootstrapServers", Type: "string", Description: "Kafka bootstrap server", Priority: 1},
```
Comment (rhuss) [inline pkg/factories/rune_factory.go:258]: Not sure if this is possbile, but I can think that the list of bootstrap servers can be quite large as a column. Is there a way to truncate this to a max. nr of chars ?

Same for topics. Also I wouldn't add _every_ field of a source as a user can always use `describe` for more details. While I think that the bootstrap servers might be important to distinguish different servers, I don't think we should include the consumergroup by default, and maybe even not the topic. 

Not sure about the Age, but this might be helpful, too (i would check whether other list commands include the age, too, as the Age is probably available everywhere)

I might even prefer to list the _sink_ instead (in a short form), as this is an important information whether a source is actively sending events, or not.

So in summary:

* Remove Topics & ConsumerGroup, shorten bootstrap serves to e.g. 30 chars. Add Age only if other sources add them, too.
* Add sink in a short form, if this possible. Again this should be kind of aligned with other sources (don't know it by heart how it looks like for the other sources though

**Review by rhuss — COMMENTED** [review]
Looks good with some minor request for changes. 

For the selected columns I would check what the other sources are offering (like: do the show the sink or age ?). Always remember, that screen real estate is limited as we only have a single line per source and we should avoid that the display becomes larger than 130 chars as this will mess up the display on most terminals.

File: pkg/factories/rune_factory.go [hunk 7]
```diff
@@ -242,3 +299,46 @@ func writeKafkaSource(dw printers.PrefixWriter, source *v1alpha1.KafkaSource) {
...
+	}
+
+	if options.AllNamespaces {
+		row.Cells = append(row.Cells, kafkaSource.ObjectMeta.Namespace)
+	}
+
+	row.Cells = append(row.Cells,
+		trunc(kafkaSource.ObjectMeta.Name),
+		commands.Age(kafkaSource.ObjectMeta.CreationTimestamp.Time),
+		trunc(kafkaSource.Spec.Sink.Ref.Name),
```
Comment (dsimansk) [inline pkg/factories/rune_factory.go:316]: Wrt other suggestion about sing prefix, that's how it's achieved in `kn`.
```suggestion
		trunc(flags.SinkToString(*kafkaSource.Spec.Sink)),
```

Comment (dsimansk) [reply pkg/factories/rune_factory.go:316]:
```
NAME         AGE   SINK                 BOOTSTRAPSERVERS
mykafkasrc   13d   ksvc:event-display   my-cluster-kafka-bootstrap.kafka.svc:9092
```

**Comment by rhuss** [comment]
thanks ! Have not much time right now to dig deeper but we should merge it right now. Please unhold if done.

/approve
/lgtm
/hold

**Comment by dsimansk** [comment]
/lgtm
Unholding wrt Roland's comment and all the changes present.
/unhold
